### PR TITLE
Misc Image Caches

### PR DESCRIPTION
**Pull request by RemieRichards**
- Caches the rock edge overlays of rock turfs - <B>Rough Change:</B> num_turfs*4 images --> 4 images
- Caches the alien weeds edging overlays - <B>Rough Change:</B> num_weeds*4 images --> 4 images
- Caches the overlays of SMES - <B>Rough Change:</B> num_smes*9 images --> 9 images

On top of the <B>Rough Changes</B> theres also the fact that these images won't constantly be recreated over and over and over again.

Will add more if I make more.
Available for theft/ports.

### CODE REVIEW

**Comment by WJohn** [comment]
What does this change, mechanically, if anything? Or is it just faster/less prone to bugs?

**Comment by RemieRichards** [comment]
@WJohn I'll explain it with the rock turfs one since it's probably the biggest improvement.

on the 255x255 map of mining, theres lets say, 200x200 turfs that are rock.
The entire edge of that 200x200 "block" will have on average 2 neighbours, and 2 open spots.
for each of those open spots every single rock turf has to create an image and place it in it's overlays list, so that's 2*number of edge turfs which I'd guess at the moment is around
2,000 images (of the 40,000 turfs that would be present in a 200x200, I'd say 1,000 if that are on the edge)
 
On top of that, everytime a mining turf is destroyed, all the other mining turfs around it check to see if they need to place or remove any of these edge images, which means creating them again.

So what it does in this case is replaces 2,000 (roundstart) images with 4 that get reused when needed.

tl;dr stops unnecessary images from being created and reuse them where possible.

**Comment by WJohn** [comment]
Huh. I wasn't aware the asteroid was so inefficient. I don't know if it causes any noticeable lag during play, but this is a nice boost when there's several miners.

**Comment by MrStonedOne** [comment]
The main improvement is in memory usage, and world start time. (not round start, mind you)

**Comment by Dorsisdwarf** [comment]
If it makes updating the server &rebooting faster, I'm all for it

**Comment by optimumtact** [comment]
generic icon caching framework when remie :tangerine:

**Comment by RemieRichards** [comment]
@optimumtact I did partway refactor this to a more generic system, but decided against it as it'd force us to have either all caches indexed by number (good luck remembering) or keep all defines defined, which slows compile time.
